### ISSUE TITLE
Workflow looses values on some tasks 

### ISSUE THREAD

**Issue by jansonz** [Netflix/conductor#1761]
I'm experiencing very weird issue where when JSONPATH extract from one task output it works on some steps but some got empty input value. For example I have several tasks in workflow pointing to output from another task like this: ${convert_ism_xml_to_json_1.output.response.json.smil.body.switch.video[*].@src}. When using this as input one one stepthe input value is there, however on another it becomes empty... Any ideas? Have triplechecked and there is no typo... It seems a bit random and doesn't occur all the time. I think mainly when there are several workflows running at once. Single Conductor installation.

**Comment by jansonz** [comment]
Adding some more examples.

The bellow is a JSON copy from a failed workflow. The task takes as input "files": "${convert_ism_xml_to_json_1.output.response.json.smil.body.switch.video[*].@src}" which as visible ends up as "files": []

`{
"taskType": "fork_generator",
"status": "COMPLETED",
"inputData": {
"variables": {
"filename_postfix": "_track1.h264",
"processing_location": "/mnt/4711ffdf-040d-408b-b09d-6db562ab0828",
"task_postfix": "add",
"files": []
},
"task_type": "mp4box-add"
},
"referenceTaskName": "fork_generator_mp4box_add_1",
"retryCount": 0,
"seq": 78,
"pollCount": 1,
"taskDefName": "fork_generator",
"scheduledTime": 1593513560349,
"startTime": 1593513560576,
"endTime": 1593513560745,
"updateTime": 1593513560576,
"startDelayInSeconds": 0,
"retried": false,
"executed": true,
"callbackFromWorker": true,
"responseTimeoutSeconds": 600,
"workflowInstanceId": "0eac86cf-baeb-4369-8316-0c41d216d37b",
"workflowType": "content_processing",
"taskId": "cec6a2ec-7ade-4cfd-8e0b-b7f376f39257",
"callbackAfterSeconds": 0,
"workerId": "Janiss-MacBook-Pro.local",
"outputData": {
"task_input": {},
"tasks": []
},
"workflowTask": {
"name": "fork_generator",
"taskReferenceName": "fork_generator_mp4box_add_1",
"inputParameters": {
"task_type": "mp4box-add",
"variables": {
"files": "${convert_ism_xml_to_json_1.output.response.json.smil.body.switch.video[*].@src}",
"filename_postfix": "_track1.h264",
"processing_location": "${workflow.input.provider_config.properties.processing_location}${workflow.input.vott_asset_id}",
"task_postfix": "add"
}
},
"type": "SIMPLE",
"decisionCases": {},
"defaultCase": [],
"forkTasks": [],
"startDelay": 0,
"joinOn": [],
"optional": false,
"taskDefinition": {
"createTime": 1592994552968,
"createdBy": "",
"name": "fork_generator",
"retryCount": 0,
"timeoutSeconds": 3600,
"inputKeys": [],
"outputKeys": [],
"timeoutPolicy": "RETRY",
"retryLogic": "FIXED",
"retryDelaySeconds": 2,
"responseTimeoutSeconds": 600,
"inputTemplate": {},
"rateLimitPerFrequency": 0,
"rateLimitFrequencyInSeconds": 1,
"ownerEmail": "xxxxxx"
},
"defaultExclusiveJoinTask": [],
"asyncComplete": false,
"loopOver": []
},
"rateLimitPerFrequency": 0,
"rateLimitFrequencyInSeconds": 0,
"workflowPriority": 0,
"iteration": 0,
"taskDefinition": {
"present": true
},
"loopOverTask": false,
"queueWaitTime": 227,
"taskStatus": "COMPLETED",
"logs": []
}`

This is not correct as it actually contains a list of files and the same ${convert_ism_xml_to_json_1.output.response.json.smil.body.switch.video[*].@src} in another step on the same workflow contains the array with data:

`"files": [
"XX_Westworld_15486_S2E8_org_V1_100.ismv",
"XX_Westworld_15486_S2E8_org_V1_200.ismv",
"XX_Westworld_15486_S2E8_org_V1_300.ismv",
"XX_Westworld_15486_S2E8_org_V1_500.ismv",
"XX_Westworld_15486_S2E8_org_V1_800.ismv",
"XX_Westworld_15486_S2E8_org_V1_1200.ismv",
"XX_Westworld_15486_S2E8_org_V1_1600.ismv",
"XX_Westworld_15486_S2E8_org_V1_2200.ismv",
"XX_Westworld_15486_S2E8_org_V1_4000.ismv",
"XX_Westworld_15486_S2E8_org_V1_6000.ismv"
],`

Full JSON:

`{
"taskType": "fork_generator",
"status": "COMPLETED",
"inputData": {
"variables": {
"vod_ing_host": "xxxx",
"processing_location": "/mnt/queued/4711ffdf-040d-408b-b09d-6db562ab0828",
"rewrite_extension": "mp4",
"file_type": "video",
"files": [
"XX_Westworld_15486_S2E8_org_V1_100.ismv",
"XX_Westworld_15486_S2E8_org_V1_200.ismv",
"XX_Westworld_15486_S2E8_org_V1_300.ismv",
"XX_Westworld_15486_S2E8_org_V1_500.ismv",
"XX_Westworld_15486_S2E8_org_V1_800.ismv",
"XX_Westworld_15486_S2E8_org_V1_1200.ismv",
"XX_Westworld_15486_S2E8_org_V1_1600.ismv",
"XX_Westworld_15486_S2E8_org_V1_2200.ismv",
"XX_Westworld_15486_S2E8_org_V1_4000.ismv",
"XX_Westworld_15486_S2E8_org_V1_6000.ismv"
],
"asset_id": xxxx
},
"task_type": "vod-ingest"
},
"referenceTaskName": "fork_generator_vod-ingest_1",
"retryCount": 0,
"seq": 115,
"pollCount": 1,
"taskDefName": "fork_generator",
"scheduledTime": 1593513581750,
"startTime": 1593513582023,
"endTime": 1593513582226,
"updateTime": 1593513582023,
"startDelayInSeconds": 0,
"retried": false,
"executed": true,
"callbackFromWorker": true,
"responseTimeoutSeconds": 600,
"workflowInstanceId": "0eac86cf-baeb-4369-8316-0c41d216d37b",
"workflowType": "content_processing",
"taskId": "2d11df84-c27c-4acd-9839-0b5dc801b568",
"callbackAfterSeconds": 0,
"workerId": "xxxxx",
"outputData": {
},
"workflowTask": {
"name": "fork_generator",
"taskReferenceName": "fork_generator_vod-ingest_1",
"inputParameters": {
"task_type": "vod-ingest",
"variables": {
"files": "${convert_ism_xml_to_json_1.output.response.json.smil.body.switch.video[*].@src}",
"processing_location": "${workflow.input.provider_config.properties.processing_location}${workflow.input.vott_asset_id}",
"asset_id": "${get_vott_asset_1.output.response.json._items[0].atv_assetid}",
"file_type": "video",
"vod_ing_host": "${get_vott_cluster_1.output.response.json.vod_ingest}",
"rewrite_extension": "mp4"
}
},
"type": "SIMPLE",
"decisionCases": {},
"defaultCase": [],
"forkTasks": [],
"startDelay": 0,
"joinOn": [],
"optional": false,
"taskDefinition": {
"createTime": 1592994552968,
"createdBy": "",
"name": "fork_generator",
"retryCount": 0,
"timeoutSeconds": 3600,
"inputKeys": [],
"outputKeys": [],
"timeoutPolicy": "RETRY",
"retryLogic": "FIXED",
"retryDelaySeconds": 2,
"responseTimeoutSeconds": 600,
"inputTemplate": {},
"rateLimitPerFrequency": 0,
"rateLimitFrequencyInSeconds": 1,
"ownerEmail": "xxxxxx"
},
"defaultExclusiveJoinTask": [],
"asyncComplete": false,
"loopOver": []
},
"rateLimitPerFrequency": 0,
"rateLimitFrequencyInSeconds": 0,
"workflowPriority": 0,
"iteration": 0,
"taskDefinition": {
"present": true
},
"loopOverTask": false,
"queueWaitTime": 273,
"taskStatus": "COMPLETED",
"logs": []
}`

What could be the reason that ${convert_ism_xml_to_json_1.output.response.json.smil.body.switch.video[*].@src} behaves differently within the same workflow? This is also not consistent - if I start the workflow again with exactly the same values it can also complete fine with all variables containing correct values.

I'm not 100% sure but this might be triggered if I have several workflows running in parallel...

**Comment by kishorebanala** [comment]
Hey @jansonz , Do you see any logs when resolving this input parameter: ```${convert_ism_xml_to_json_1.output.response.json.smil.body.switch.video[*].@src}``` ?

Also, do you have locking enabled on this instance? You can use LocalOnlyLock introduced in this PR to rule out issues due to race conditions when evaluating workflows: https://github.com/Netflix/conductor/pull/1728

If you still see issues, consider debugging the issue around JSONPath evaluation logic that resolves above path. Alternatively, you can avoid expressions within JSON Path by extracting the output with in previous task, and passing the input to next tasks as a simple path.. eg: {convert_ism_xml_to_json_1.output.files}. But, I'd recommend first approach to see if there are any inherent bugs in JSON Path evaluations.

**Comment by jansonz** [comment]
Thank you for the feedback.

Have added workflow.dyno.queue.sharding.strategy=localOnly to the Conductor configuration and will test.

This is a single conductor server installation so not sure if this will make any difference, but fingers crossed.

**Comment by jansonz** [comment]
@kishorebanala No success. Still ran in the same issue and it seems to affect these kinds of JSONPATH expressions: ${convert_ism_xml_to_json_1.output.response.json.smil.body.switch.audio[*].@src} and ${convert_ism_xml_to_json_1.output.response.json.smil.body.switch.video[*].@src} which should return a list of files. some tasks in the same workflow end up with value and some don't. have triple checked typos but as this sometimes works and sometimes doesn't based on the same input it's quite unlikely :( 
Can you elaborate a bit on "you can avoid expressions within JSON Path by extracting the output with in previous task, and passing the input to next tasks as a simple path"? Do you mean I should write worked which could do this or there is a more simple way?

**Comment by kishorebanala** [comment]
@jansonz Well, one option is to use a Lambda task in between (which also uses Nashorn javascript engine). Otherwise, you can do this output transformation as part of a simple task inside your worker.

We've plans to migrate away to better javascript rendering engines like [JQ](https://stedolan.github.io/jq/) in the future, which would make it easier to write and extract complex json path expressions.
